db/transaction.go

```diff
@@ -1,89 +1,128 @@
+// consistency.
 type Transaction struct {
-	colInfo *colInfo
-	txn     *leveldb.Transaction
+	mut             sync.Mutex
+	colInfo         *colInfo
+	batchWriter     dbBatchWriter
+	readerWithBatch *readerWithBatchWriter
+	committed       bool
+	discarded       bool
 }
 
-// OpenTransaction opens an atomic DB transaction. Only one transaction can be
-// opened at a time. All write operations (e.g. Insert, Update, or Delete) will
-// be blocked until in-flight transaction is committed or discarded. The
-// returned transaction is safe for concurrent use.
+// OpenTransaction opens and returns a new transaction. While the transaction is
+// open, no other state changes (e.g. Insert, Update, or Delete) can be made to
+// the database (but concurrent reads are still allowed).
+//
+// Transactions are atomic, meaning that either:
 //
-// Transaction is expensive and can overwhelm compaction, especially if
-// transaction size is small. Use with caution.
+//     (1) The transaction will succeed and *all* queued operations will be
+//     applied, or
+//     (2) the transaction will be fail or be discarded, in which case *none* of
+//     the queued operations will be applied.
 //
 // The transaction must be closed once done, either by committing or discarding
-// the transaction. Closing the DB will discard any open transactions.
-func (c *Collection) OpenTransaction() (*Transaction, error) {
-	txn, err := c.ldb.OpenTransaction()
-	if err != nil {
-		return nil, err
-	}
+// the transaction. No changes will be made to the database state until the
+// transaction is committed.
+func (c *Collection) OpenTransaction() *Transaction {
+	c.info.writeMut.Lock()
 	return &Transaction{
-		colInfo: c.info.copy(),
-		txn:     txn,
-	}, nil
+		colInfo:         c.info.copy(),
+		batchWriter:     c.ldb,
+		readerWithBatch: newReaderWithBatchWriter(c.ldb),
+	}
+}
+
+// checkState acquires a lock on txn.mut and then calls unsafeCheckState.
+func (txn *Transaction) checkState() error {
+	txn.mut.Lock()
+	defer txn.mut.Unlock()
+	return txn.unsafeCheckState()
+}
+
+// unsafeCheckState checks the state of the transaction, assuming the caller has
+// already acquired a lock. It returns an error if the transaction has already
+// been committed or discarded.
+func (txn *Transaction) unsafeCheckState() error {
+	if txn.discarded {
+		return ErrDiscarded
+	} else if txn.committed {
+		return ErrCommitted
+	}
+	return nil
 }
 
 // Commit commits the transaction. If error is not nil, then the transaction is
 // not committed, it can then either be retried or discarded.
 //
 // Other methods should not be called after transaction has been committed.
 func (txn *Transaction) Commit() error {
-	return txn.txn.Commit()
+	txn.mut.Lock()
+	if err := txn.unsafeCheckState(); err != nil {
+		txn.mut.Unlock()
+		return err
+	}
+	txn.committed = true
+	txn.mut.Unlock()
+	defer txn.colInfo.writeMut.Unlock()
+	return txn.batchWriter.Write(txn.readerWithBatch.batch, nil)
```

Comment:
You write after releasing the locks?

Reply:
There are two mutexes. `colInfo.writeMux` is a per-collection global write lock. txn.mut` just protects the `committed` and `discarded` properties of the transaction. I thought it was okay set `committed = true` and release `txn.mut` before we actually write, but this is technically not correct. I updated the code so that `committed = true` is not set and the write lock is not released unless the commit was successful.